### PR TITLE
[VectorCombine] Add support for zext/sext/trunc to shuffleToIdentity

### PR DESCRIPTION
**Pull request by davemgreen**
This is one of the simple additions to shuffleToIdentity that help it look through intermediate zext/sext instructions.

### CODE REVIEW

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll [hunk 1]
```diff
@@ -432,19 +432,10 @@ define <8 x half> @fma(<8 x half> %a, <8 x half> %b, <8 x half> %c) {
...
-; CHECK-NEXT:    [[AB1:%.*]] = zext <4 x i32> [[AB]] to <4 x i64>
-; CHECK-NEXT:    [[AT1:%.*]] = zext <4 x i32> [[AT]] to <4 x i64>
-; CHECK-NEXT:    [[BB1:%.*]] = sext <4 x i32> [[BB]] to <4 x i64>
-; CHECK-NEXT:    [[BT1:%.*]] = sext <4 x i32> [[BT]] to <4 x i64>
-; CHECK-NEXT:    [[ABB:%.*]] = add <4 x i64> [[AB1]], [[BB1]]
-; CHECK-NEXT:    [[ABT:%.*]] = add <4 x i64> [[AT1]], [[BT1]]
-; CHECK-NEXT:    [[ABB1:%.*]] = trunc <4 x i64> [[ABB]] to <4 x i32>
-; CHECK-NEXT:    [[ABT1:%.*]] = trunc <4 x i64> [[ABT]] to <4 x i32>
-; CHECK-NEXT:    [[R:%.*]] = shufflevector <4 x i32> [[ABB1]], <4 x i32> [[ABT1]], <8 x i32> <i32 0, i32 4, i32 1, i32 5, i32 2, i32 6, i32 3, i32 7>
+; CHECK-NEXT:    [[TMP1:%.*]] = zext <8 x i32> [[A:%.*]] to <8 x i64>
```
Comment (artagnon) [inline llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:435]: I think there is value in this transform, because #92668 can't do anything like this: it can't generate new zext instructions with a different number of elements.

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 2]
```diff
@@ -1773,7 +1772,9 @@ bool VectorCombine::foldShuffleToIdentity(Instruction &I) {
         !cast<BinaryOperator>(Item[0].first)->isIntDivRem()) {
       Worklist.push_back(GenerateInstLaneVectorFromOperand(Item, 0));
       Worklist.push_back(GenerateInstLaneVectorFromOperand(Item, 1));
-    } else if (isa<UnaryOperator>(Item[0].first)) {
+    } else if (isa<UnaryOperator>(Item[0].first) ||
```
Comment (artagnon) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1775]: Use match() from PatternMatch?

**Review by artagnon — COMMENTED** [review]
Hi,

I did a partial review of the existing code, and my suggestions are along the lines of the following diff. I think it would be nice if you could improve the existing code as an NFC patch before landing this patch.

```diff
diff --git a/llvm/lib/Transforms/Vectorize/VectorCombine.cpp b/llvm/lib/Transforms/Vectorize/VectorCombine.cpp
index 8573a8adf53b..52830d7ed462 100644
--- a/llvm/lib/Transforms/Vectorize/VectorCombine.cpp
+++ b/llvm/lib/Transforms/Vectorize/VectorCombine.cpp
@@ -1686,7 +1686,7 @@ bool VectorCombine::foldShuffleToIdentity(Instruction &I) {
       int M = SV->getMaskValue(Lane);
       if (M < 0)
         return {nullptr, PoisonMaskElem};
-      else if (M < (int)NumElts) {
+      if (static_cast<unsigned>(M) < NumElts) {
         V = SV->getOperand(0);
         Lane = M;
       } else {
@@ -1700,12 +1700,12 @@ bool VectorCombine::foldShuffleToIdentity(Instruction &I) {
   auto GenerateInstLaneVectorFromOperand =
       [&LookThroughShuffles](ArrayRef<InstLane> Item, int Op) {
         SmallVector<InstLane> NItem;
-        for (InstLane V : Item) {
-          NItem.emplace_back(
-              !V.first
-                  ? InstLane{nullptr, PoisonMaskElem}
-                  : LookThroughShuffles(
-                        cast<Instruction>(V.first)->getOperand(Op), V.second));
+        for (InstLane IL : Item) {
+          auto [V, Lane] = IL;
+          // The caller is responsible for ensuring that this is always an Instruction.
+          auto *I = cast<Instruction>(V);
+          InstLane ToInsert = V ? LookThroughShuffles(I->getOperand(Op), Lane) : IL;
+          NItem.emplace_back(ToInsert);
         }
         return NItem;
       };
@@ -1721,57 +1721,66 @@ bool VectorCombine::foldShuffleToIdentity(Instruction &I) {

   while (!Worklist.empty()) {
     SmallVector<InstLane> Item = Worklist.pop_back_val();
-    if (++NumVisited > MaxInstrsToScan)
+    if (NumVisited++ > MaxInstrsToScan)
       return false;

+    auto [FrontV, FrontLane] = Item.front();
+
     // If we found an undef first lane then bail out to keep things simple.
-    if (!Item[0].first)
+    if (!FrontV)
       return false;

     // Look for an identity value.
-    if (Item[0].second == 0 && Item[0].first->getType() == Ty &&
-        all_of(drop_begin(enumerate(Item)), [&](const auto &E) {
-          return !E.value().first || (E.value().first == Item[0].first &&
-                                      E.value().second == (int)E.index());
+    if (!FrontLane && FrontV->getType() == Ty &&
+        all_of(drop_begin(enumerate(Item)), [Item](const auto &I) {
+          auto [FrontV, _] = Item.front();
+          auto [Idx, E] = I;
+          auto [V, Lane] = E;
+          return !V || (V == FrontV && Lane == (int)Idx);
         })) {
-      IdentityLeafs.insert(Item[0].first);
+      IdentityLeafs.insert(FrontV);
       continue;
     }
+
     // Look for a splat value.
-    if (all_of(drop_begin(Item), [&](InstLane &IL) {
-          return !IL.first ||
-                 (IL.first == Item[0].first && IL.second == Item[0].second);
+    if (all_of(drop_begin(Item), [Item](InstLane &IL) {
+          auto [FrontV, FrontLane] = Item.front();
+          auto [V, Lane] = IL;
+          return !V ||
+                 (V == FrontV && Lane == FrontLane);
         })) {
-      SplatLeafs.insert(Item[0].first);
+      SplatLeafs.insert(FrontV);
       continue;
     }

     // We need each element to be the same type of value, and check that each
     // element has a single use.
-    if (!all_of(drop_begin(Item), [&](InstLane IL) {
-          if (!IL.first)
+    if (!all_of(drop_begin(Item), [Item](InstLane IL) {
+          auto [FrontV, _] = Item.front();
+          auto [V, _Lane] = IL;
+          if (!V)
             return true;
-          if (auto *I = dyn_cast<Instruction>(IL.first); I && !I->hasOneUse())
+          if (auto *I = dyn_cast<Instruction>(V); I && !I->hasOneUse())
             return false;
-          if (IL.first->getValueID() != Item[0].first->getValueID())
+          if (V->getValueID() != FrontV->getValueID())
             return false;
-          if (isa<CallInst>(IL.first) && !isa<IntrinsicInst>(IL.first))
+          if (isa<CallInst>(V) && !isa<IntrinsicInst>(V))
             return false;
-          auto *II = dyn_cast<IntrinsicInst>(IL.first);
+          auto *II = dyn_cast<IntrinsicInst>(V);
           return !II ||
-                 (isa<IntrinsicInst>(Item[0].first) &&
+                 (isa<IntrinsicInst>(FrontV) &&
                   II->getIntrinsicID() ==
-                      cast<IntrinsicInst>(Item[0].first)->getIntrinsicID());
+                      cast<IntrinsicInst>(FrontV)->getIntrinsicID());
         }))
       return false;

     // Check the operator is one that we support. We exclude div/rem in case
     // they hit UB from poison lanes.
-    if (isa<BinaryOperator>(Item[0].first) &&
-        !cast<BinaryOperator>(Item[0].first)->isIntDivRem()) {
+    if (isa<BinaryOperator>(FrontV) &&
+        !cast<BinaryOperator>(FrontV)->isIntDivRem()) {
       Worklist.push_back(GenerateInstLaneVectorFromOperand(Item, 0));
       Worklist.push_back(GenerateInstLaneVectorFromOperand(Item, 1));
-    } else if (isa<UnaryOperator>(Item[0].first)) {
+    } else if (isa<UnaryOperator>(FrontV)) {
       Worklist.push_back(GenerateInstLaneVectorFromOperand(Item, 0));
     } else {
       return false;
@@ -1790,7 +1799,7 @@ bool VectorCombine::foldShuffleToIdentity(Instruction &I) {
       return Item[0].first;
     }
     if (SplatLeafs.contains(Item[0].first)) {
-      if (auto ILI = dyn_cast<Instruction>(Item[0].first))
+      if (auto *ILI = dyn_cast<Instruction>(Item[0].first))
         Builder.SetInsertPoint(*ILI->getInsertionPointAfterDef());
       else if (isa<Argument>(Item[0].first))
         Builder.SetInsertPointPastAllocas(I.getParent()->getParent());
@@ -1803,7 +1812,7 @@ bool VectorCombine::foldShuffleToIdentity(Instruction &I) {
     for (unsigned Idx = 0, E = I->getNumOperands(); Idx < E; Idx++)
       Ops[Idx] = Generate(GenerateInstLaneVectorFromOperand(Item, Idx));
     Builder.SetInsertPoint(I);
-    if (auto BI = dyn_cast<BinaryOperator>(I))
+    if (auto *BI = dyn_cast<BinaryOperator>(I))
       return Builder.CreateBinOp((Instruction::BinaryOps)BI->getOpcode(),
                                  Ops[0], Ops[1]);
     assert(isa<UnaryInstruction>(I) &&
--
2.45.1
```

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 3]
```diff
@@ -1773,7 +1772,9 @@ bool VectorCombine::foldShuffleToIdentity(Instruction &I) {
...
+               isa<TruncInst>(Item[0].first) || isa<ZExtInst>(Item[0].first) ||
+               isa<SExtInst>(Item[0].first)) {
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1777]:
```suggestion
    } else if (isa<UnaryOperator,TruncInst,ZExtInst,SExtInst>(Item[0].first))
```

**Comment by davemgreen** [comment]
> I did a partial review of the existing code, and my suggestions are along the lines of the following diff. I think it would be nice if you could improve the existing code as an NFC patch before landing this patch.

Do you want to put together a patch for them? It is hard to see why the cast<Instruction> is always correct but some of the other changes look decent.

**Comment by artagnon** [comment]
> Do you want to put together a patch for them? It is hard to see why the cast is always correct but some of the other changes look decent.

Sure, I'll also try to simplify the algorithm.

**Comment by RKSimon** [comment]
rebase?

**Comment by davemgreen** [comment]
Yeah will do.

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll [hunk 4]
```diff
@@ -15,9 +15,7 @@ define <8 x i8> @trivial(<8 x i8> %a) {
 
 define <4 x i32> @add_same_operands(<4 x i32> %x) {
 ; CHECK-LABEL: @add_same_operands(
-; CHECK-NEXT:    [[SHUF:%.*]] = shufflevector <4 x i32> [[X:%.*]], <4 x i32> poison, <4 x i32> <i32 3, i32 2, i32 1, i32 0>
-; CHECK-NEXT:    [[ADD:%.*]] = add <4 x i32> [[SHUF]], [[SHUF]]
-; CHECK-NEXT:    [[REVSHUF:%.*]] = shufflevector <4 x i32> [[ADD]], <4 x i32> poison, <4 x i32> <i32 3, i32 2, i32 1, i32 0>
+; CHECK-NEXT:    [[REVSHUF:%.*]] = add <4 x i32> [[X:%.*]], [[X]]
```
Comment (artagnon) [inline llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:18]: Why did this test get updated? Is it a consequence of loosening Operand0 and Operand1 to be anything, not just Instructions?

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll [hunk 5]
```diff
@@ -372,8 +370,7 @@ define <8 x i8> @inner_shuffle(<8 x i8> %a, <8 x i8> %b, <8 x i8> %c) {
 define <4 x i32> @extrause_add_same_operands(<4 x i32> %x) {
 ; CHECK-LABEL: @extrause_add_same_operands(
 ; CHECK-NEXT:    [[SHUF:%.*]] = shufflevector <4 x i32> [[X:%.*]], <4 x i32> poison, <4 x i32> <i32 3, i32 2, i32 1, i32 0>
-; CHECK-NEXT:    [[ADD:%.*]] = add <4 x i32> [[SHUF]], [[SHUF]]
-; CHECK-NEXT:    [[REVSHUF:%.*]] = shufflevector <4 x i32> [[ADD]], <4 x i32> poison, <4 x i32> <i32 3, i32 2, i32 1, i32 0>
+; CHECK-NEXT:    [[REVSHUF:%.*]] = add <4 x i32> [[X]], [[X]]
```
Comment (artagnon) [inline llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:373]: Not a correctness issue, but this variable shouldn't be named `REVSHUF`. Same for the one below.

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll [hunk 6]
```diff
@@ -600,19 +595,10 @@ define <8 x i16> @not_bitcast2(<4 x i32> %x, <8 x i16> %y) {
...
-; CHECK-NEXT:    [[AB1:%.*]] = zext <4 x i32> [[AB]] to <4 x i64>
-; CHECK-NEXT:    [[AT1:%.*]] = zext <4 x i32> [[AT]] to <4 x i64>
-; CHECK-NEXT:    [[BB1:%.*]] = sext <4 x i32> [[BB]] to <4 x i64>
-; CHECK-NEXT:    [[BT1:%.*]] = sext <4 x i32> [[BT]] to <4 x i64>
-; CHECK-NEXT:    [[ABB:%.*]] = add <4 x i64> [[AB1]], [[BB1]]
-; CHECK-NEXT:    [[ABT:%.*]] = add <4 x i64> [[AT1]], [[BT1]]
-; CHECK-NEXT:    [[ABB1:%.*]] = trunc <4 x i64> [[ABB]] to <4 x i32>
-; CHECK-NEXT:    [[ABT1:%.*]] = trunc <4 x i64> [[ABT]] to <4 x i32>
-; CHECK-NEXT:    [[R:%.*]] = shufflevector <4 x i32> [[ABB1]], <4 x i32> [[ABT1]], <8 x i32> <i32 0, i32 4, i32 1, i32 5, i32 2, i32 6, i32 3, i32 7>
+; CHECK-NEXT:    [[TMP1:%.*]] = zext <8 x i32> [[A:%.*]] to <8 x i64>
```
Comment (artagnon) [inline llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:598]: Would be nice if we could name this variable `ZEXT`.

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll [hunk 7]
```diff
@@ -754,10 +720,8 @@ define void @trunc(<8 x i64> %a, <8 x i64> %b, ptr %p) {
 
 define <4 x i64> @zext_chain(<4 x i16> %x) {
```
Comment (artagnon) [inline llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:721]: Why didn't `add_chain` and `zext_add_chain` get updated?

**Review by artagnon — COMMENTED** [review]
There seem to be some underlying issues, not directly related to your patch.

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll (hunk 5 above)
Comment (artagnon) [reply llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:373]: Unless we're creating new instructions, in which case, variable names are a nice-to-have, I don't think we should worry about variable names.

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll (hunk 7 above)
Comment (artagnon) [reply llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:721]: Can you kindly verify this? I'm mainly worried about some underlying bug in the algorithm, which could result in correctness issues as opposed to missed optimizations. It seems very strange to me that a single add, but not a chain of adds, is handled. If multiple uses aren't handled, even a single add with identical operands shouldn't be reduced. If a chain of adds and chain of zexts are handled, a combination of the two should be handled automatically. At least that's how my (now abandoned) algorithm worked.

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 8]
```diff
@@ -1753,8 +1756,7 @@ static Value *generateNewInstTree(ArrayRef<InstLane> Item, FixedVectorType *Ty,
 // do so.
 bool VectorCombine::foldShuffleToIdentity(Instruction &I) {
   auto *Ty = dyn_cast<FixedVectorType>(I.getType());
-  if (!Ty || !isa<Instruction>(I.getOperand(0)) ||
-      !isa<Instruction>(I.getOperand(1)))
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1757]: Isn't this an unrelated change?

Comment (davemgreen) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1757]: Sure - I can split it out. It will be needed for many of the tests, but there should be enough to test various cases still.

File: llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll (hunk 7 above)
Comment (davemgreen) [reply llvm/test/Transforms/VectorCombine/AArch64/shuffletoidentity.ll:721]: Yeah It should be correct, if not optimal. The algorithm needs to check for extra uses to make sure that there will not introduce more instructions. This isn't true of the highest shuffle though, which does not have the same one-use check requirement.

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 9]
```diff
@@ -1742,6 +1742,9 @@ static Value *generateNewInstTree(ArrayRef<InstLane> Item, FixedVectorType *Ty,
   if (auto *BI = dyn_cast<BinaryOperator>(I))
     return Builder.CreateBinOp((Instruction::BinaryOps)BI->getOpcode(), Ops[0],
                                Ops[1]);
+  if (auto CI = dyn_cast<CastInst>(I))
```
[the remark this answers is not in the document]
Comment (artagnon) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1749]: After this nit is fixed, this patch LGTM.

**Review by artagnon — APPROVED** [review]
